src/website/shared/logs/logs.py

```diff
@@ -100,95 +72,155 @@ def _annotate_username(self, query: EventQuerySet) -> EventQuerySet:
+                pgh_created_at=Subquery(
+                    DerivationClusterProposalLinkEvent.objects.filter(
+                        proposal_id=OuterRef("proposal_id")
                     )
+                    .order_by("pgh_created_at")
+                    .values("pgh_created_at")[:1]
                 )
-                .filter(proposal_id__in=suggestion_ids)
-                .annotate(dummy_group_by_value=Value(1))
-                .values("dummy_group_by_value")
-                # NOTE(alejandrosame): Doing annotate "dummy_group_by_value" is a hack shared in
-                # the Django forum. Another user there complains about Django's decision to add the
-                # primary key by default when doing automated groupings. We hit here the same problem.
-                # Reference: https://forum.djangoproject.com/t/excess-group-by-columns-causing-problems-with-window-functions/26865/1
             )
-            .annotate(
-                suggestion_id=Cast(F("proposal_id"), BigIntegerField()),
-                timestamp=F("pgh_created_at"),
-                action=F("pgh_label"),
-                status_value=Value("NOT_A_STATUS_CHANGE"),
-                package_names=ArrayAgg(
-                    JSONObject(
-                        name="derivation__name",
-                        attribute=Replace(
-                            "derivation__attribute",  # type: ignore
-                            Concat(Value("."), "derivation__system"),  # type: ignore
-                            Value(""),
-                        ),
+            .filter(proposal_id__in=suggestion_ids)
+        ).annotate(
+            package_names=ArrayAgg(
+                JSONObject(
+                    name="derivation__name",
+                    attribute=Replace(
+                        "derivation__attribute",  # type: ignore
+                        Concat(Value("."), "derivation__system"),  # type: ignore
+                        Value(""),
                     ),
-                    distinct=True,
                 ),
-                package_count=Count("derivation__name", distinct=True),
-            )
-            .values(*fields)
+                distinct=True,
+            ),
+            package_count=Count("derivation__name", distinct=True),
         )
 
-        return status_qs.union(package_qs).order_by("timestamp")
+        for package_event in package_qs.all().iterator():
+            raw_events.append(
+                {
+                    "suggestion_id": package_event.proposal_id,
+                    "timestamp": package_event.pgh_created_at,
+                    "username": package_event.username,
+                    "action": package_event.pgh_label,
+                    "package_names": package_event.package_names,
+                    "package_count": package_event.package_count,
+                }
+            )
+
+        return sorted(raw_events, key=lambda event: event["timestamp"])
 
     def get_dict(
```

Comment:
The schema should simply be e.g. a Pydantic model, which will provide a `model_dump()` method out of the box. Then we don't need any comments to document how it's supposed to work.

Reply:
Ah, great idea, I just forgot that we have unions in Python :sweat_smile: